packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/ReactViewGroup.java

```diff
@@ -1083,6 +1063,136 @@ public void setBackfaceVisibilityDependantOpacity() {
+          }
+          return super.performAccessibilityAction(host, action, args);
+        }
+        return super.performAccessibilityAction(host, action, args);
+      }
+    };
+    ViewCompat.setAccessibilityDelegate(this, viewAccessibilityDelegate);
+    this.setFocusable(true);
+    this.setFocusableInTouchMode(true);
+    // To force Talkback to give the a11y event to the FocusGuide
+    // we need to make it look like it has some a11y label to be announced.
+    // Because FocusGuide should always have at least one focusable child view,
+    // which will receive forwarded a11y event from this FocusGuide,
+    // the following fake label will never be announced
+    this.setContentDescription("FocusGuide");
+    this.isFocusGuideTalkbackAccessibilityDelegateSet = true;
+  }
+
+  public void cleanupFocusGuideTalkbackAccessibilityDelegate() {
+    ViewCompat.setAccessibilityDelegate(this, null);
+    this.setFocusable(this.originalIsFocusable);
+    this.setFocusableInTouchMode(this.originalIsFocusable);
+    this.originalIsFocusable = false;
+    this.setContentDescription(null);
+    this.isFocusGuideTalkbackAccessibilityDelegateSet = false;
+  }
+
+  public boolean hasFocusGuideTalkbackAccessibilityDelegate() {
+    return this.isFocusGuideTalkbackAccessibilityDelegateSet;
+  }
+
+  private boolean interceptAccessibilityEvents(int action, Bundle args) {
+    if (!this.isTVFocusGuide()) {
+      return false;
+    }
+
+    // If it's "FocusGuide", we want to intercept
+    // the event and redirect it to either:
+    // 1) first available destination view from "destinations" prop
+    // 2) last focused element saved when "autoFocus" prop is `true`
+    // 3) first focusable child view
+    View destinationView = this.findDestinationView();
+    if (destinationView != null) {
+      try {
+        destinationView.performAccessibilityAction(action, args);
+        return true;
+      } catch (Exception e) {
+        FLog.e(TAG, "Exception when performing accessibility action on destination view - falling back to next case (last focused view): " + e);
+      }
+    }
+    View lastFocusedView = this.lastFocusedElement.get();
+    if (lastFocusedView != null) {
+      try {
+        lastFocusedView.performAccessibilityAction(action, args);
+        return true;
+      } catch (Exception e) {
+        FLog.e(TAG, "Exception when performing accessibility action on last focused view - falling back to next case (first focusable view): " + e);
+      }
+    }
+    View firstFocusableView = this.getFirstFocusableView(this);
+    if (firstFocusableView != null) {
+      try {
+        firstFocusableView.performAccessibilityAction(action, args);
+        return true;
+      } catch (Exception e) {
+        FLog.e(TAG, "Exception when performing accessibility action on first focusable view - focus guide will not handle focus: " + e);
+      }
+    }
+    return false;
+  }
+
+  private View getFirstFocusableView(ReactViewGroup viewGroup) {
```

Comment:
If you could move this method back up to around line 511 (before `moveFocusToFirstFocusable()`) the PR would result in a smaller diff and be easier to review and cherry pick.

Reply:
That will result in a duplicated code, between those two methods - how about moving focus guide methods/utilities to a new file (e.g. FocusGuideUtil.java/.kt) so that ReactViewGroup patches are easier to maintain?

Reply:
Your current implementation is fine -- I was just asking that the `getFirstFocusableView` method start at line 511 instead of at the current location (line 1161).

Reply:
Oh sorry, I didn't read your comment carefully enough lol - fixed